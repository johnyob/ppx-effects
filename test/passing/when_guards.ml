exception%effect Foo : bool -> unit

let _ =
  match false with
  | _ when true -> 'a'
  | _ -> 'b'
  | exception Not_found -> 'c'
  | exception Failure s when String.equal s "Oops" -> 'd'
  | [%effect? Foo b, _] when b -> 'e'
  | [%effect? Foo b, _ when not b] -> 'e'
  | [%effect? Foo _, k] -> EffectHandlers.Deep.continue k ()
;;
