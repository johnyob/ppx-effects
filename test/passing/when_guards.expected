open! Ppx_effects_runtime.Import
type _ Ppx_effects_runtime.t +=  
  | Foo: bool -> unit Ppx_effects_runtime.t 
let _ =
  Ppx_effects_runtime.Deep.match_with (fun `Unit -> false) `Unit
    {
      Ppx_effects_runtime.Deep.retc =
        (function | _ when true -> 'a' | _ -> 'b');
      exnc =
        (function
         | Not_found -> 'c'
         | Failure s when String.equal s "Oops" -> 'd'
         | e -> Ppx_effects_runtime.raise e);
      effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function
              | Foo b when b -> Some ((fun _ -> 'e'))
              | Foo b when not b -> Some ((fun _ -> 'e'))
              | Foo _ -> Some ((fun k -> continue k ()))
              | _ -> None : continue_input Ppx_effects_runtime.t ->
                              ((continue_input, _)
                                 Ppx_effects_runtime.Deep.continuation -> 
                                 _)
                                option) in
         effc)
    }
